### DISCUSSION TITLE
AQI calculations in HA

### DISCUSSION THREAD

**Discussion by ByteTheHand** [siku2/hass-weatherlink#36] [Show and tell]
First off, I want to say a huge thank you to siku2 for developing this integration.  I've been running it for several weeks now and it is rock solid and has allowed me to create automations based off of the data.  If you have a Patreon, please let me know.  I'd support the effort with code, but not really up to speed on coding these days.

Second, if people would like a conversion from the raw PM 2.5 data to the US AQI numbers, based on the same calculation that you see on your Davis WeatherLink page let me know.  I implemented this through creating a sensor that returns the calculated AQI number.  Since I modified my version to run off of the 1-minute PM value, this gives me the "real-time" AQI that I use to turn on/off window fans base on current outdoor air conditions.

While AQI is formally a backwards looking 24-hour calculation that is built on PM 2.5, PM10 and Ozone readings, that doesn't really help much if you want to run automations, thus the reliance on "real-time AQI".  You'll find the real-time calculation on the Davis WeatherLink page and on Purple Air.  Very useful in quickly changing conditions.

**Comment by ardyp87** [comment]
This would be great to have!

**Comment by ByteTheHand** [comment]
Here is the code for my outdoor AirLink:

...

template:
  - sensor:
      - name: "Outdoor_AQI"
        unit_of_measurement: "AQI"
        state: >
          {% set out_pm_2_5 = states("sensor.airlink_pm_2_5") | float %}
          
          {% if out_pm_2_5 <= 12.0 %}
            {{ (4.166667 * out_pm_2_5) | int }}
          {% elif out_pm_2_5 <= 35.4 %}
            {{ ((2.103004 * (out_pm_2_5-12.1)) + 51) | int }}
          {% elif out_pm_2_5 <= 55.4 %}
            {{ ((2.462312 * (out_pm_2_5-35.5)) + 101) | int }}
          {% elif out_pm_2_5 <= 150.4 %}
            {{ ((0.516333 * (out_pm_2_5-55.5)) + 151) | int }}
          {% elif out_pm_2_5 <= 250.4 %}
            {{ ((0.990991 * (out_pm_2_5-150.5)) + 201) | int }}
          {% elif out_pm_2_5 <= 500.4 %}
            {{ ((0.796319 * (out_pm_2_5-250.5)) + 301) | int }}
          {% else %}
            {{ (out_pm_2_5 + 500) | int }}
          {% endif %}
        unique_id: outdoor_aqi 

...

Pasted into your configuration.yaml it will return a US EPA AirNow AQI value.

Note:  The other change I made was to the actual python code in one module to return the 1 minute AQI rather than the NowCast AQI.  in the sensor_air_quality.py file.  You can do that by looking for the following line (130):

…
    @property
    def state(self):
        return self.units.pm.convert(self._aq_condition.pm_2p5)
/#        return self.units.pm.convert(self._aq_condition.pm_2p5_nowcast)

…

I commented out the _nowcast value and returned the current value instead.  Once this goes to production I will likely lose that capability.
